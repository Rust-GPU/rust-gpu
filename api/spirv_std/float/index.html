<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Traits and helper functions related to floats."><title>spirv_std::float - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-42caa33d.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="spirv_std" data-themes="" data-resource-suffix="" data-rustdoc-version="1.84.0-nightly (b19329a37 2024-11-21)" data-channel="nightly" data-search-js="search-92e6798f.js" data-settings-js="settings-0f613d39.js" ><script src="../../static.files/storage-59e33391.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-5f194d8c.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../spirv_std/index.html">spirv_<wbr>std</a><span class="version">0.9.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module float</a></h2><h3><a href="#traits">Module Items</a></h3><ul class="block"><li><a href="#traits" title="Traits">Traits</a></li><li><a href="#functions" title="Functions">Functions</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate spirv_<wbr>std</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><span class="rustdoc-breadcrumbs"><a href="../index.html">spirv_std</a></span><h1>Module <span>float</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/spirv_std/float.rs.html#1-240">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Traits and helper functions related to floats.</p>
</div></details><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">ยง</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.Float.html" title="trait spirv_std::float::Float">Float</a></div><div class="desc docblock-short">Abstract trait representing a SPIR-V floating point type.</div></li></ul><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">ยง</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.f16_to_f32.html" title="fn spirv_std::float::f16_to_f32">f16_<wbr>to_<wbr>f32</a></div><div class="desc docblock-short">Converts an f16 (half) into an f32 (float). The parameter is a u32, due to GPU support for u16
not being universal - the upper 16 bits are ignored.</div></li><li><div class="item-name"><a class="fn" href="fn.f16x2_to_vec2.html" title="fn spirv_std::float::f16x2_to_vec2">f16x2_<wbr>to_<wbr>vec2</a></div><div class="desc docblock-short">Converts two f16 values (halfs) into two f32 values (floats). The parameter is a u32, with the
low 16 bits being the first f16, and the high 16 bits being the second f16.</div></li><li><div class="item-name"><a class="fn" href="fn.f32_to_f16.html" title="fn spirv_std::float::f32_to_f16">f32_<wbr>to_<wbr>f16</a></div><div class="desc docblock-short">Converts an f32 (float) into an f16 (half). The result is a u32, not a u16, due to GPU support
for u16 not being universal - the upper 16 bits will always be zero.</div></li><li><div class="item-name"><a class="fn" href="fn.u8x4_to_vec4_snorm.html" title="fn spirv_std::float::u8x4_to_vec4_snorm">u8x4_<wbr>to_<wbr>vec4_<wbr>snorm</a></div><div class="desc docblock-short">Unpacks 4 8-bit signed integers into a vec4. See
<a href="https://www.khronos.org/registry/SPIR-V/specs/1.0/GLSL.std.450.html">UnpackSnorm4x8</a> for exact
semantics.</div></li><li><div class="item-name"><a class="fn" href="fn.u8x4_to_vec4_unorm.html" title="fn spirv_std::float::u8x4_to_vec4_unorm">u8x4_<wbr>to_<wbr>vec4_<wbr>unorm</a></div><div class="desc docblock-short">Unpacks 4 8-bit unsigned integers into a vec4. See
<a href="https://www.khronos.org/registry/SPIR-V/specs/1.0/GLSL.std.450.html">UnpackSnorm4x8</a> for exact
semantics.</div></li><li><div class="item-name"><a class="fn" href="fn.u16x2_to_vec2_snorm.html" title="fn spirv_std::float::u16x2_to_vec2_snorm">u16x2_<wbr>to_<wbr>vec2_<wbr>snorm</a></div><div class="desc docblock-short">Unpacks 2 16-bit signed integers into a vec2. See
<a href="https://www.khronos.org/registry/SPIR-V/specs/1.0/GLSL.std.450.html">UnpackSnorm2x16</a> for
exact semantics.</div></li><li><div class="item-name"><a class="fn" href="fn.u16x2_to_vec2_unorm.html" title="fn spirv_std::float::u16x2_to_vec2_unorm">u16x2_<wbr>to_<wbr>vec2_<wbr>unorm</a></div><div class="desc docblock-short">Unpacks 2 16-bit unsigned integers into a vec2. See
<a href="https://www.khronos.org/registry/SPIR-V/specs/1.0/GLSL.std.450.html">UnpackUnorm2x16</a> for
exact semantics.</div></li><li><div class="item-name"><a class="fn" href="fn.vec2_to_f16x2.html" title="fn spirv_std::float::vec2_to_f16x2">vec2_<wbr>to_<wbr>f16x2</a></div><div class="desc docblock-short">Converts two f32 values (floats) into two f16 values (halfs). The result is a u32, with the low
16 bits being the first f16, and the high 16 bits being the second f16.</div></li><li><div class="item-name"><a class="fn" href="fn.vec2_to_u16x2_snorm.html" title="fn spirv_std::float::vec2_to_u16x2_snorm">vec2_<wbr>to_<wbr>u16x2_<wbr>snorm</a></div><div class="desc docblock-short">Packs a vec2 into 2 16-bit signed integers. See
<a href="https://www.khronos.org/registry/SPIR-V/specs/1.0/GLSL.std.450.html">PackSnorm2x16</a> for exact
semantics.</div></li><li><div class="item-name"><a class="fn" href="fn.vec2_to_u16x2_unorm.html" title="fn spirv_std::float::vec2_to_u16x2_unorm">vec2_<wbr>to_<wbr>u16x2_<wbr>unorm</a></div><div class="desc docblock-short">Packs a vec2 into 2 16-bit unsigned integers. See
<a href="https://www.khronos.org/registry/SPIR-V/specs/1.0/GLSL.std.450.html">PackUnorm2x16</a> for exact
semantics.</div></li><li><div class="item-name"><a class="fn" href="fn.vec4_to_u8x4_snorm.html" title="fn spirv_std::float::vec4_to_u8x4_snorm">vec4_<wbr>to_<wbr>u8x4_<wbr>snorm</a></div><div class="desc docblock-short">Packs a vec4 into 4 8-bit signed integers. See
<a href="https://www.khronos.org/registry/SPIR-V/specs/1.0/GLSL.std.450.html">PackSnorm4x8</a> for exact
semantics.</div></li><li><div class="item-name"><a class="fn" href="fn.vec4_to_u8x4_unorm.html" title="fn spirv_std::float::vec4_to_u8x4_unorm">vec4_<wbr>to_<wbr>u8x4_<wbr>unorm</a></div><div class="desc docblock-short">Packs a vec4 into 4 8-bit unsigned integers. See
<a href="https://www.khronos.org/registry/SPIR-V/specs/1.0/GLSL.std.450.html">PackUnorm4x8</a> for exact
semantics.</div></li></ul></section></div></main></body></html>