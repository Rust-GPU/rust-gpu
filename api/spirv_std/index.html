<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="`spirv-std`"><title>spirv_std - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="spirv_std" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (244da22fa 2024-04-23)" data-channel="nightly" data-search-js="search-bf21c90c8c1d92b1.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-e32f0c247825364d.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../spirv_std/index.html">spirv_std</a><span class="version">0.9.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../spirv_std/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">spirv_std</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/spirv_std/lib.rs.html#1-152">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h2 id="spirv-std"><a class="doc-anchor" href="#spirv-std">§</a><code>spirv-std</code></h2>
<p>Core functions, traits, and more that make up a “standard library” for SPIR-V for use in <a href="https://github.com/rust-gpu/rust-gpu#readme">rust-gpu</a>.</p>
<p>This crate gives a <code>rust-gpu</code> shader access to the required <code>#![spirv(..)]</code> attribute, as well as povide all kinds of APIs that allows a shader to access GPU resources such as textures and buffers. Optionally, through the use of the <code>&quot;glam&quot;</code> feature, it includes some boilerplate trait implementations to make <code>glam</code> vector types compatible with these APIs.</p>
<h3 id="example"><a class="doc-anchor" href="#example">§</a>Example</h3>
<p><img src="https://github.com/rust-gpu/rust-gpu/raw/b12a2f3f6a54bc841d05a9224bc577909d519228/docs/assets/sky.jpg" alt="Sky shader" /></p>
<p>Here is a small excerpt to see what a shader would look like. See <a href="https://github.com/rust-gpu/rust-gpu/blob/69cb69d28f1e64420ee31ade5e7dffb7c5621e89/examples/shaders/sky-shader/src/lib.rs">source</a> for full details of the shader that generates above image.</p>

<div class="example-wrap ignore"><a href="#" class="tooltip" title="This example is not tested">ⓘ</a><pre class="rust rust-example-rendered"><code><span class="kw">use </span>spirv_std::spirv;
<span class="kw">use </span>glam::{Vec3, Vec4, vec2, vec3};

<span class="attr">#[spirv(fragment)]
</span><span class="kw">pub fn </span>main(
    <span class="attr">#[spirv(frag_coord)] </span>in_frag_coord: <span class="kw-2">&amp;</span>Vec4,
    <span class="attr">#[spirv(push_constant)] </span>constants: <span class="kw-2">&amp;</span>ShaderConstants,
    output: <span class="kw-2">&amp;mut </span>Vec4,
) {
    <span class="kw">let </span>frag_coord = vec2(in_frag_coord.x, in_frag_coord.y);
    <span class="kw">let </span><span class="kw-2">mut </span>uv = (frag_coord - <span class="number">0.5 </span>* vec2(constants.width <span class="kw">as </span>f32, constants.height <span class="kw">as </span>f32))
        / constants.height <span class="kw">as </span>f32;
    uv.y = -uv.y;

    <span class="kw">let </span>eye_pos = vec3(<span class="number">0.0</span>, <span class="number">0.0997</span>, <span class="number">0.2</span>);
    <span class="kw">let </span>sun_pos = vec3(<span class="number">0.0</span>, <span class="number">75.0</span>, -<span class="number">1000.0</span>);
    <span class="kw">let </span>dir = get_ray_dir(uv, eye_pos, sun_pos);

    <span class="comment">// evaluate Preetham sky model
    </span><span class="kw">let </span>color = sky(dir, sun_pos);

    <span class="kw-2">*</span>output = tonemap(color).extend(<span class="number">1.0</span>)
}</code></pre></div>
<h3 id="getting-started"><a class="doc-anchor" href="#getting-started">§</a>Getting started</h3>
<p>Check out <a href="https://rust-gpu.github.io/rust-gpu/book/">The <code>rust-gpu</code> Dev Guide</a> for information on how to get started with using it in your projects.</p>
<p>Experiment with rust-gpu shaders in-browser at <a href="https://shadered.org/shaders?language=rust&amp;sort=hot">SHADERed</a>.</p>
</div></details><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><code>pub extern crate <a class="mod" href="../spirv_std_macros/index.html" title="mod spirv_std_macros">spirv_std_macros</a> as macros;</code></div></li><li><div class="item-name" id="reexport.ByteAddressableBuffer"><code>pub use byte_addressable_buffer::<a class="struct" href="byte_addressable_buffer/struct.ByteAddressableBuffer.html" title="struct spirv_std::byte_addressable_buffer::ByteAddressableBuffer">ByteAddressableBuffer</a>;</code></div></li><li><div class="item-name" id="reexport.num_traits"><code>pub use <a class="mod" href="../num_traits/index.html" title="mod num_traits">num_traits</a>;</code></div></li><li><div class="item-name" id="reexport.glam"><code>pub use <a class="mod" href="../glam/index.html" title="mod glam">glam</a>;</code></div></li></ul><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="arch/index.html" title="mod spirv_std::arch">arch</a></div><div class="desc docblock-short">SPIR-V Intrinsics</div></li><li><div class="item-name"><a class="mod" href="byte_addressable_buffer/index.html" title="mod spirv_std::byte_addressable_buffer">byte_addressable_buffer</a></div><div class="desc docblock-short">Container for an untyped blob of data.</div></li><li><div class="item-name"><a class="mod" href="float/index.html" title="mod spirv_std::float">float</a></div><div class="desc docblock-short">Traits and helper functions related to floats.</div></li><li><div class="item-name"><a class="mod" href="image/index.html" title="mod spirv_std::image">image</a></div><div class="desc docblock-short">Image types</div></li><li><div class="item-name"><a class="mod" href="indirect_command/index.html" title="mod spirv_std::indirect_command">indirect_command</a></div><div class="desc docblock-short">Indirect command structs from vulkan</div></li><li><div class="item-name"><a class="mod" href="integer/index.html" title="mod spirv_std::integer">integer</a></div><div class="desc docblock-short">Traits related to integers.</div></li><li><div class="item-name"><a class="mod" href="memory/index.html" title="mod spirv_std::memory">memory</a></div><div class="desc docblock-short">Types for handling memory ordering constraints for concurrent memory access.</div></li><li><div class="item-name"><a class="mod" href="number/index.html" title="mod spirv_std::number">number</a></div><div class="desc docblock-short">Traits and helper functions related to numbers.</div></li><li><div class="item-name"><a class="mod" href="ray_tracing/index.html" title="mod spirv_std::ray_tracing">ray_tracing</a></div><div class="desc docblock-short">Ray-tracing data types</div></li><li><div class="item-name"><a class="mod" href="scalar/index.html" title="mod spirv_std::scalar">scalar</a></div><div class="desc docblock-short">Traits related to scalars.</div></li><li><div class="item-name"><a class="mod" href="vector/index.html" title="mod spirv_std::vector">vector</a></div><div class="desc docblock-short">Traits related to vectors.</div></li></ul><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.Image.html" title="macro spirv_std::Image">Image</a></div><div class="desc docblock-short">A macro for creating SPIR-V <code>OpTypeImage</code> types. Always produces a
<code>spirv_std::image::Image&lt;...&gt;</code> type.</div></li><li><div class="item-name"><a class="macro" href="macro.ray_query.html" title="macro spirv_std::ray_query">ray_query</a></div><div class="desc docblock-short">Constructs an uninitialized ray query variable. Using the syntax
<code>let (mut)? &lt;name&gt;</code>. Where <code>name</code> is the name of the ray query variable.</div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.RuntimeArray.html" title="struct spirv_std::RuntimeArray">RuntimeArray</a></div><div class="desc docblock-short">Dynamically-sized arrays in Rust carry around their length as the second half of a tuple.
Unfortunately, sometimes SPIR-V provides an unsized array with no way of obtaining its length.
Hence, this type represents something very similar to a slice, but with no way of knowing its
length.</div></li><li><div class="item-name"><a class="struct" href="struct.Sampler.html" title="struct spirv_std::Sampler">Sampler</a></div><div class="desc docblock-short">An opaque reference to settings that describe how to access, filter, or
sample an image.</div></li><li><div class="item-name"><a class="struct" href="struct.TypedBuffer.html" title="struct spirv_std::TypedBuffer">TypedBuffer</a></div><div class="desc docblock-short">Explicit (uniform/storage) buffer handle for descriptor indexing.</div></li></ul><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.spirv.html" title="attr spirv_std::spirv">spirv</a></div><div class="desc docblock-short">Replaces all (nested) occurrences of the <code>#[spirv(..)]</code> attribute with
<code>#[cfg_attr(target_arch=&quot;spirv&quot;, rust_gpu::spirv(..))]</code>.</div></li></ul></section></div></main></body></html>