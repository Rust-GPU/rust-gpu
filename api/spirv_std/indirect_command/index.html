<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Indirect command structs from vulkan"><title>spirv_std::indirect_command - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-42caa33d.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="spirv_std" data-themes="" data-resource-suffix="" data-rustdoc-version="1.84.0-nightly (b19329a37 2024-11-21)" data-channel="nightly" data-search-js="search-92e6798f.js" data-settings-js="settings-0f613d39.js" ><script src="../../static.files/storage-59e33391.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-5f194d8c.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../spirv_std/index.html">spirv_<wbr>std</a><span class="version">0.9.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module indirect_<wbr>command</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate spirv_<wbr>std</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><span class="rustdoc-breadcrumbs"><a href="../index.html">spirv_std</a></span><h1>Module <span>indirect_command</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/spirv_std/indirect_command.rs.html#1-187">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Indirect command structs from vulkan</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.DispatchIndirectCommand.html" title="struct spirv_std::indirect_command::DispatchIndirectCommand">Dispatch<wbr>Indirect<wbr>Command</a></div><div class="desc docblock-short">Structure specifying an indirect dispatching command</div></li><li><div class="item-name"><a class="struct" href="struct.DrawIndexedIndirectCommand.html" title="struct spirv_std::indirect_command::DrawIndexedIndirectCommand">Draw<wbr>Indexed<wbr>Indirect<wbr>Command</a></div><div class="desc docblock-short">Structure specifying an indexed indirect drawing command</div></li><li><div class="item-name"><a class="struct" href="struct.DrawIndirectCommand.html" title="struct spirv_std::indirect_command::DrawIndirectCommand">Draw<wbr>Indirect<wbr>Command</a></div><div class="desc docblock-short">Structure specifying an indirect drawing command</div></li><li><div class="item-name"><a class="struct" href="struct.DrawMeshTasksIndirectCommandEXT.html" title="struct spirv_std::indirect_command::DrawMeshTasksIndirectCommandEXT">Draw<wbr>Mesh<wbr>Tasks<wbr>Indirect<wbr>CommandEXT</a></div><div class="desc docblock-short">Structure specifying a mesh tasks draw indirect command</div></li><li><div class="item-name"><a class="struct" href="struct.TraceRaysIndirectCommand2KHR.html" title="struct spirv_std::indirect_command::TraceRaysIndirectCommand2KHR">Trace<wbr>Rays<wbr>Indirect<wbr>Command2KHR</a></div><div class="desc docblock-short">Structure specifying the parameters of an indirect trace ray command with indirect shader binding tables</div></li><li><div class="item-name"><a class="struct" href="struct.TraceRaysIndirectCommandKHR.html" title="struct spirv_std::indirect_command::TraceRaysIndirectCommandKHR">Trace<wbr>Rays<wbr>Indirect<wbr>CommandKHR</a></div><div class="desc docblock-short">Structure specifying the parameters of an indirect ray tracing command</div></li></ul><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">ยง</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.DeviceAddress.html" title="type spirv_std::indirect_command::DeviceAddress">Device<wbr>Address</a></div><div class="desc docblock-short"><a href="https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkDeviceAddress.html">https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkDeviceAddress.html</a></div></li><li><div class="item-name"><a class="type" href="type.DeviceSize.html" title="type spirv_std::indirect_command::DeviceSize">Device<wbr>Size</a></div><div class="desc docblock-short"><a href="https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkDeviceSize.html">https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkDeviceSize.html</a></div></li></ul></section></div></main></body></html>