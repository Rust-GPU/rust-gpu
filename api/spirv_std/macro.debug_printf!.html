<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.debug_printf.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.debug_printf.html">macro.debug_printf.html</a>...</p>
    <script>location.replace("macro.debug_printf.html" + location.search + location.hash);</script>
</body>
</html>