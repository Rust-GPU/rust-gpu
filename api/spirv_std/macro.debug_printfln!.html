<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.debug_printfln.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.debug_printfln.html">macro.debug_printfln.html</a>...</p>
    <script>location.replace("macro.debug_printfln.html" + location.search + location.hash);</script>
</body>
</html>