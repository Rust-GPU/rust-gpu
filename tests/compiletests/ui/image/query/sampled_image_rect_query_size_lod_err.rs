// build-fail
// normalize-stderr-test "\S*/crates/spirv-std/src/" -> "$SPIRV_STD_SRC/"
// compile-flags: -C target-feature=+ImageQuery,+SampledRect
// ignore-vulkan1.0
// ignore-vulkan1.1
// ignore-vulkan1.1spv1.4
// ignore-vulkan1.2

use spirv_std::{Image, arch, image::SampledImage, spirv};

#[spirv(fragment)]
pub fn main(
    #[spirv(descriptor_set = 0, binding = 0)] rect_sampled: &SampledImage<
        Image!(rect, type=f32, sampled),
    >,
    output: &mut glam::UVec2,
) {
    // This should fail because rect images don't support query_size_lod
    *output = rect_sampled.query_size_lod(0);
}
