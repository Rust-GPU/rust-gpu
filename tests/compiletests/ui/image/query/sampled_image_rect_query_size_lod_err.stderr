error[E0277]: the trait bound `Image<f32, 4, 2, 0, 0, 1, 0, 4>: HasQuerySizeLod` is not satisfied
    --> $DIR/sampled_image_rect_query_size_lod_err.rs:19:28
     |
19   |     *output = rect_sampled.query_size_lod(0);
     |                            ^^^^^^^^^^^^^^ the trait `HasQuerySizeLod` is not implemented for `Image<f32, 4, 2, 0, 0, 1, 0, 4>`
     |
     = help: the following other types implement trait `HasQuerySizeLod`:
               Image<SampledType, 0, DEPTH, ARRAYED, 0, SAMPLED, FORMAT, COMPONENTS>
               Image<SampledType, 1, DEPTH, ARRAYED, 0, SAMPLED, FORMAT, COMPONENTS>
               Image<SampledType, 2, DEPTH, ARRAYED, 0, SAMPLED, FORMAT, COMPONENTS>
               Image<SampledType, 3, DEPTH, ARRAYED, 0, SAMPLED, FORMAT, COMPONENTS>
note: required by a bound in `SampledImage::<Image<SampledType, DIM, DEPTH, ARRAYED, spirv_std::::image::{impl#9}::{constant#0}, SAMPLED, FORMAT, COMPONENTS>>::query_size_lod`
    --> /image.rs:1138:12
     |
1124 |     pub fn query_size_lod<Size: ImageSizeQuery<u32, DIM, ARRAYED> + Default>(
     |            -------------- required by a bound in this associated function
...
1138 |         >: HasQuerySizeLod,
     |            ^^^^^^^^^^^^^^^ required by this bound in `SampledImage::<Image<SampledType, DIM, DEPTH, ARRAYED, spirv_std::::image::{impl#9}::{constant#0}, SAMPLED, FORMAT, COMPONENTS>>::query_size_lod`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
